Add JIRA_Issue in related fields of Finding

RomainJufer opened the pull request:
This PR add the `JIRA_Issue` associated with a finding to the FindingSerializer, taking advantage of the `related_fields` field. It also add support for this new field in the swagger doc.

**Note** : We already have the fields `jira_creation` and `jira_change` which are part of the `JIRA_Issue` model so we might want to remove them and only have them in the `related_fields` ?

madchap commented on dojo/api_v2/serializers.py:
```diff
@@ -674,14 +674,22 @@ class Meta:
 
 
-class FindingTestSerializer(serializers.ModelSerializer):
+class FindingReatedFieldsSerializer(serializers.ModelSerializer):
```
Surely a typo?

RomainJufer replied:
Yes you're right, my bad, thank you

madchap commented:
This looks good to me. Regarding the note, I will let @valentijnscholten as he's done the refactoring not so long ago and surely has an opinion :-) Thanks for the PR.

valentijnscholten commented:
I left the jira_change and jira_creation fields on the Finding response for backwards compatibility.
Do we want to only return the JIRA issue key? Or the full url? Or both?

RomainJufer commented:
I fixed the typo, thank you for noticing and sorry :). Also I'll let the fields for backaward compatibility as you mentioned. About the question you asked @valentijnscholten I don't know if you want me to weigh in but currently I'm reusing an existing serializer so if we need to filter/process jira fields it would most likely mean adding a new serializer just for that, if that matters ?

RomainJufer commented:
We also noticed that in `JIRA_Issue`, both `jira_creation` and `jira_change` are strings but the same fields in the top level of Finding are annotated to be `array[string]`. Is it normal ? When we generate the client based on the doc it fails when unmarshalling these fields.

valentijnscholten commented:
> I fixed the typo, thank you for noticing and sorry :). Also I'll let the fields for backaward compatibility as you mentioned. About the question you asked @valentijnscholten I don't know if you want me to weigh in but currently I'm reusing an existing serializer so if we need to filter/process jira fields it would most likely mean adding a new serializer just for that, if that matters ?

I was thinking that if we just get the jira issue key it is not of much use if you don't know the url? Could we add the full url to the issue as a readonly field to the existing serializer?

valentijnscholten commented:
> We also noticed that in `JIRA_Issue`, both `jira_creation` and `jira_change` are strings but the same fields in the top level of Finding are annotated to be `array[string]`. Is it normal ? When we generate the client based on the doc it fails when unmarshalling these fields.

Looks like some poor copy/pasting from my side in a previous PR. Could you check if it can be easily fixed?

RomainJufer commented:
Okay, so I had to do a few changes.
- I corrected the swagger for both `jira_creation` and `jira_change`
- I added the url as you suggested inside the `JiraIssueSerializer`
- Since I added the new `jira` field inside `related_fields` for coherence I also had to reorganize that field a bit. All the fields that were top level are now under `test` as they should. Otherwise if we don't do that we will have, for example, an `id` field in top level without any information about its meaning.

Let me know what you think and thanks for your feedback :)

valentijnscholten commented on dojo/api_v2/serializers.py:
```diff
@@ -529,13 +529,25 @@ def validate(self, data):
+            return None
 
+        jira_instance = jira_project.jira_instance
+        return jira_instance.url
```
I would like to see the full url of the issue wdyt?

valentijnscholten commented:
not sure when it was introduced, but I am seeing errors on engagement reports: #3426. maybe this PR would be a good time to check/fix that?

RomainJufer commented:
Oh sorry, I'm not very experience with Jira so I did not totally understand what you meant by full URL to the issue. It should be correct now :). I also fixed the issue #3426. I simply check that the request exists in the context. I don't really know why sometime it does not exists since it is provided by a `context_processor`. Anyway, it should be fine now.

valentijnscholten reviewed (changes requested):
can't you just use `jira_helper.get_jira_issue_url` to prevent another place where jira urls are constructed?

RomainJufer commented:
If I'm not mistaken the function you mentioned expect an object which is not directly the `JIRA_Issue` but an object that reference one (hence the check `has_jira_issue`). On the other hand, in the `JIRAIssueSerializer` we have directly a `JIRA_Issue` so the `if` in the function you mentioned will fail. Also, I don't think I could use `get_jira_project_url` either for similar reason.

valentijnscholten commented:
The code in `get_url` is almost identical as `get_jira_project_url` in the jira helper. We should stop scattering jira related logic around the project and keep it in the jira helper.

RomainJufer commented:
I agree with you, I've updated the helper so that the function `get_jira_issue_url` acts on an issue and not a finding. I also updated all the call made to that function. Let me know if you think that there are other things to change.

valentijnscholten commented on dojo/jira_link/helper.py:
```diff
@@ -131,34 +131,35 @@ def get_jira_url(obj):
-            # example http://jira.com/browser/SEC-123
-            return jira_project.jira_instance.url + '/browse/' + obj.jira_issue.jira_key
+    jira_project = issue.jira_project
+    if jira_project is None:
+        return None
 
-    return None
+    jira_instance = jira_project.jira_instance
```
please leave the call to `get_jira_project` and `get_jira_instance` in place instead of accessing properties deirectly.

RomainJufer commented:
I added back the call to the function you mentioned but I've got a problem with a weird merge which resulted in all the files committed again. I'll fix that, ignore last commit for the moment

valentijnscholten commented:
resolved a conflict

RomainJufer commented:
Do you know why the tests are falling ? It seems to be something from the import-scan endpoint which should be altered by the changes in this PR

valentijnscholten commented:
first error in the logs is this one:
```
2020-12-17T15:29:15.4390584Z [36muwsgi_1         |[0m ERROR:django.request:Internal Server Error: /api/v2/findings/
2020-12-17T15:29:15.4391561Z [36muwsgi_1         |[0m Traceback (most recent call last):
2020-12-17T15:29:15.4392546Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
2020-12-17T15:29:15.4393520Z [36muwsgi_1         |[0m     response = get_response(request)
2020-12-17T15:29:15.4394622Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
2020-12-17T15:29:15.4395893Z [36muwsgi_1         |[0m     response = self.process_exception_by_middleware(e, request)
2020-12-17T15:29:15.4397115Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
2020-12-17T15:29:15.4398321Z [36muwsgi_1         |[0m     response = wrapped_callback(request, *callback_args, **callback_kwargs)
2020-12-17T15:29:15.4399559Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
2020-12-17T15:29:15.4400574Z [36muwsgi_1         |[0m     return view_func(*args, **kwargs)
2020-12-17T15:29:15.4401597Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/viewsets.py", line 125, in view
2020-12-17T15:29:15.4402664Z [36muwsgi_1         |[0m     return self.dispatch(request, *args, **kwargs)
2020-12-17T15:29:15.4403762Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 509, in dispatch
2020-12-17T15:29:15.4404816Z [36muwsgi_1         |[0m     response = self.handle_exception(exc)
2020-12-17T15:29:15.4405933Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 469, in handle_exception
2020-12-17T15:29:15.4407033Z [36muwsgi_1         |[0m     self.raise_uncaught_exception(exc)
2020-12-17T15:29:15.4408266Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
2020-12-17T15:29:15.4409277Z [36muwsgi_1         |[0m     raise exc
2020-12-17T15:29:15.4410319Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 506, in dispatch
2020-12-17T15:29:15.4411735Z [36muwsgi_1         |[0m     response = handler(request, *args, **kwargs)
2020-12-17T15:29:15.4412932Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/mixins.py", line 19, in create
2020-12-17T15:29:15.4414018Z [36muwsgi_1         |[0m     self.perform_create(serializer)
2020-12-17T15:29:15.4415225Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/mixins.py", line 24, in perform_create
2020-12-17T15:29:15.4416268Z [36muwsgi_1         |[0m     serializer.save()
2020-12-17T15:29:15.4417405Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/rest_framework/serializers.py", line 205, in save
2020-12-17T15:29:15.4418570Z [36muwsgi_1         |[0m     self.instance = self.create(validated_data)
2020-12-17T15:29:15.4419704Z [36muwsgi_1         |[0m   File "/app/dojo/api_v2/serializers.py", line 898, in create
2020-12-17T15:29:15.4420829Z [36muwsgi_1         |[0m     new_finding.save(push_to_jira=push_to_jira)
2020-12-17T15:29:15.4421742Z [36muwsgi_1         |[0m   File "/app/dojo/models.py", line 2192, in save
2020-12-17T15:29:15.4422585Z [36muwsgi_1         |[0m     jira_helper.push_to_jira(self)
2020-12-17T15:29:15.4423522Z [36muwsgi_1         |[0m   File "/app/dojo/jira_link/helper.py", line 375, in push_to_jira
2020-12-17T15:29:15.4424514Z [36muwsgi_1         |[0m     return add_jira_issue(finding)
2020-12-17T15:29:15.4425416Z [36muwsgi_1         |[0m   File "/app/dojo/decorators.py", line 61, in __wrapper__
2020-12-17T15:29:15.4426299Z [36muwsgi_1         |[0m     return func(*args, **kwargs)
2020-12-17T15:29:15.4427189Z [36muwsgi_1         |[0m   File "/app/dojo/decorators.py", line 34, in __wrapper__
2020-12-17T15:29:15.4428061Z [36muwsgi_1         |[0m     return func(*args, **kwargs)
2020-12-17T15:29:15.4429084Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/celery/local.py", line 191, in __call__
2020-12-17T15:29:15.4430131Z [36muwsgi_1         |[0m     return self._get_current_object()(*a, **kw)
2020-12-17T15:29:15.4431214Z [36muwsgi_1         |[0m   File "/usr/local/lib/python3.6/site-packages/celery/app/task.py", line 393, in __call__
2020-12-17T15:29:15.4432227Z [36muwsgi_1         |[0m     return self.run(*args, **kwargs)
2020-12-17T15:29:15.4433144Z [36muwsgi_1         |[0m   File "/app/dojo/decorators.py", line 106, in __wrapper__
2020-12-17T15:29:15.4434298Z [36muwsgi_1         |[0m     return func(*args, **kwargs)
2020-12-17T15:29:15.4435140Z [36muwsgi_1         |[0m   File "/app/dojo/jira_link/helper.py", line 484, in add_jira_issue
2020-12-17T15:29:15.4435998Z [36muwsgi_1         |[0m     jira_issue_url = get_jira_issue_url(find)
2020-12-17T15:29:15.4436887Z [36muwsgi_1         |[0m   File "/app/dojo/jira_link/helper.py", line 159, in get_jira_issue_url
2020-12-17T15:29:15.4438434Z [36muwsgi_1         |[0m     return jira_instance.url + '/browse/' + issue.jira_key
2020-12-17T15:29:15.4442799Z [36muwsgi_1         |[0m AttributeError: 'Finding' object has no attribute 'jira_key'
2020-12-17T15:29:15.4443556Z [36muwsgi_1         |[0m ERROR
```
